File: Either.py
class Either(object):
   def __init__(self, value, is_right):
     self.is_right = is_right
     self.value = value

   @staticmethod
   def right(value):
     return Either(value, True)

   @staticmethod
   def left(value):
     return Either(value, False)

   def map(self, mapping):
     if self.is_right:
       return Either.right(mapping(self.value))
     else:
       return Either.left(self.value)

   def bind(self, binding):
     if self.is_right:
       return binding(self.value)
     else:
       return Either(self.value, False)
       
   def __str__(self):
     if self.is_right:
       return "Right " + str(self.value)
     else:
       return "Left " + str(self.value)
       
right_two = Either.right(2)
print(right_two)

right_four = right_two.map(lambda x: x + 2)
print(right_four)

def divide (x,y):
  if y == 0:
    return Either.left("Cannot divide by 0")
  else:
    return Either.right(x / y)

right_one = right_four.bind(lambda x: divide(x, 4))
print(right_one)

div_by_zero = right_four.bind(lambda x: divide(x, 0))
print(div_by_zero)

File: Either.cs
class Either<Right, Left>
{
  readonly Right right;
  readonly Left left;
  readonly bool isRight = true;

  public Either(Right val)
  {
    right = val;    
  }

  public Either(Left val)
  {
    left = val;
    isRight = false;
  }
      
  public Either<T, Left> Map<T>(Func<Right, T> mapping)
  {
    if (isRight)
    {
      var result = mapping(right);
      return new Either<T, Left>(result);
    }
    return new Either<T, Left>(left);
  }

  public Either<T, Left> Bind<T>(Func<Right, Either<T, Left>> binding)
  {
    return isRight ? binding(right) : new Either<T, Left>(left);
  }

  public override string ToString()
  {
    return isRight
      ? "Right(" + right.ToString() + ")"
      : "Left(" + left.ToString() + ")";
  }
}

var five = new Either<int, string>(2)
  .Map(x => x + 3)
  .ToString();
//Right(5)

Either<int, string> Divide(int numerator, int denomenator)
{
  return denomenator == 0
    ? new Either<int, string>("cannot divide by 0")
    : new Either<int, string>(num / den);
}

var divByZero = new Either<int, string>(10)
  .Bind(x => Divide(x, 0))
  .ToString();
//Left(cannot divide by 0)

var fiveAgain = new Either<int, string>(10)
  .Bind(x => Divide(x, 2))
  .ToString();
//Right(5)
